File: c_explore/blackjack.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to calculate the total value of a hand
int calculateHandValue(int hand[], int numCards) {
    int value = 0;
    int numAces = 0;

    for (int i = 0; i < numCards; i++) {
        if (hand[i] == 1) { // Ace
            numAces++;
        }
        value += hand[i];
    }

    while (numAces > 0 && value + 10 <= 21) {
        value += 10;
        numAces--;
    }

    return value;
}

// Function to display a hand
void displayHand(int hand[], int numCards) {
    for (int i = 0; i < numCards; i++) {
        printf("%d ", hand[i]);
    }
    printf("\n");
}

int main() {
    int deck[52] = {2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10,  // Cards 2 to 10 and face cards (J, Q, K) have value 10
                    10, 10, 10, 11,                       // Ace can be 1 or 11
                    2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10,
                    10, 10, 10, 11,
                    2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10,
                    10, 10, 10, 11};

    int playerHand[10];
    int dealerHand[10];
    int numPlayerCards = 0;
    int numDealerCards = 0;

    srand(time(NULL));

    // Initial deal: Player gets two cards, Dealer gets one card face-up
    playerHand[numPlayerCards++] = deck[rand() % 52];
    playerHand[numPlayerCards++] = deck[rand() % 52];
    dealerHand[numDealerCards++] = deck[rand() % 52];

    printf("Welcome to Blackjack!\n\n");
    printf("Your hand: ");
    displayHand(playerHand, numPlayerCards);
    printf("Dealer's hand: %d\n\n", dealerHand[0]);

    // Player's turn
    while (1) {
        int choice;
        printf("Enter 1 to hit or 2 to stand: ");
        scanf("%d", &choice);

        if (choice == 1) {
            playerHand[numPlayerCards++] = deck[rand() % 52];
            printf("Your hand: ");
            displayHand(playerHand, numPlayerCards);

            int playerValue = calculateHandValue(playerHand, numPlayerCards);
            if (playerValue > 21) {
                printf("Bust! You lose.\n");
                return 0;
            }
        } else if (choice == 2) {
            break;
        } else {
            printf("Invalid choice. Please enter 1 to hit or 2 to stand.\n");
        }
    }

    // Dealer's turn
    printf("\nDealer's hand: ");
    displayHand(dealerHand, numDealerCards);
    while (1) {
        int dealerValue = calculateHandValue(dealerHand, numDealerCards);
        if (dealerValue >= 17) {
            break;
        }

        dealerHand[numDealerCards++] = deck[rand() % 52];
        printf("Dealer's hand: ");
        displayHand(dealerHand, numDealerCards);
    }

    // Determine the winner
    int playerValue = calculateHandValue(playerHand, numPlayerCards);
    int dealerValue = calculateHandValue(dealerHand, numDealerCards);

    printf("\nYour hand value: %d\n", playerValue);
    printf("Dealer's hand value: %d\n\n", dealerValue);

    if (playerValue > dealerValue && playerValue <= 21) {
        printf("Congratulations! You win!\n");
    } else if (dealerValue > 21 && playerValue <= 21) {
        printf("Dealer busts! You win!\n");
    } else {
        printf("Dealer wins. Better luck next time!\n");
    }

    return 0;
}

File: jsprojects/blackjack-game/blackjack.js
// blackjack.js
const readline = require('readline-sync');

// Function to generate a deck of cards
function generateDeck() {
  const suits = ['♠', '♣', '♦', '♥'];
  const values = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];

  const deck = [];
  for (const suit of suits) {
    for (const value of values) {
      deck.push({ suit, value });
    }
  }

  return deck;
}

// Function to shuffle the deck
function shuffle(deck) {
  for (let i = deck.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [deck[i], deck[j]] = [deck[j], deck[i]];
  }
}

// Function to calculate the total value of a hand
function getHandValue(hand) {
  let total = 0;
  let hasAce = false;

  for (const card of hand) {
    if (card.value === 'A') {
      hasAce = true;
      total += 11;
    } else if (['K', 'Q', 'J'].includes(card.value)) {
      total += 10;
    } else {
      total += parseInt(card.value, 10);
    }
  }

  if (hasAce && total > 21) {
    total -= 10;
  }

  return total;
}

// Function to display a hand
function displayHand(hand) {
  return hand.map((card) => card.suit + card.value).join(', ');
}

// Main game function
function playBlackjack() {
  const deck = generateDeck();
  shuffle(deck);

  const playerHand = [];
  const dealerHand = [];

  // Deal two cards to the player and dealer
  playerHand.push(deck.pop());
  dealerHand.push(deck.pop());
  playerHand.push(deck.pop());
  dealerHand.push(deck.pop());

  console.log('Your hand:', displayHand(playerHand));
  console.log('Dealer\'s hand:', displayHand([dealerHand[0], { suit: '??', value: '??' }]));

  while (true) {
    const playerValue = getHandValue(playerHand);
    if (playerValue === 21) {
      console.log('Blackjack! You win!');
      break;
    } else if (playerValue > 21) {
      console.log('Busted! You lose.');
      break;
    }

    const action = readline.question('Do you want to "hit" or "stand"? ');
    if (action.toLowerCase() === 'hit') {
      playerHand.push(deck.pop());
      console.log('Your hand:', displayHand(playerHand));
    } else if (action.toLowerCase() === 'stand') {
      // Dealer's turn
      const dealerValue = getHandValue(dealerHand);

      while (dealerValue < 17) {
        dealerHand.push(deck.pop());
        dealerValue = getHandValue(dealerHand);
      }

      console.log('Your hand:', displayHand(playerHand));
      console.log('Dealer\'s hand:', displayHand(dealerHand));

      if (dealerValue > 21 || dealerValue < playerValue) {
        console.log('You win!');
      } else if (dealerValue === playerValue) {
        console.log('Push! It\'s a tie.');
      } else {
        console.log('You lose.');
      }

      break;
    } else {
      console.log('Invalid input. Please enter "hit" or "stand".');
    }
  }
}

playBlackjack();

